### PR TITLE
Improve cluster resilience and move to using templated yaml files

### PR DESCRIPTION
**Pull request by MatMeredith**
This PR covers
- Adding pre-stop scripts to Chronos and Astaire so that you can scale the clusters up and down without breaking them
- Moving to generating K8s yaml files from tmpl files to reduce code management pain

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -108,7 +108,8 @@ Instead of using Docker Compose, you can deploy Clearwater in Kubernetes. This r
 
 - Update the Kubernetes yaml to match your deployment.
 
-  - In each file ending depl.yaml you will need to replace {{REPO}} with the path to the repository that you pushed your images to
+  - Generate the K8s yaml files from the templates by running `IMAGE_PATH=<path to your repo> IMAGE_TAG=<tag for the images you want to use> k8s-subst kubernetes`
```
Comment (johadalin) [inline README.md]: I think we should be consistent with the line above, referring to "Kubernetes yaml" rather than having one be that, and the other be "K8s". 
Also, the command should probably be `...k8s-subst ~/clearawter-docker/kubernetes/` to make it clear its the path to the yaml file template directory. At first reading i wasn't really clear what the `kubernetes` part of this was referring to

File: README.md [hunk 2]
```diff
@@ -37,6 +37,14 @@ By default Docker images built using clearwater-docker will pull the latest stab
 
 ![alt text](docs/images/clearwater-docker_in_scope.jpg "Clearwater-Docker dispayed in Scope")
 
+## Limitations
+
+Note that scaling of Docker deployments is a work in progress and there are currently a number of known issues...
+
+* Homestead-prov and Ellis don’t load balance across multiple Cassandra nodes.
```
Comment (johadalin) [inline README.md:44]: This is sad :( 
Is it because of the way python creates the connections that it only happens at start of day or something?

File: README.md [hunk 3]
```diff
@@ -37,6 +37,14 @@ By default Docker images built using clearwater-docker will pull the latest stab
...
+
+* In general deleting pods from storage clusters (Cassandra, Chronos or Astaire) is not supported.  Pods that are deleted will not get removed from the clusters and the clusters will end up broken.  The exception is when deployed under Kubernetes -- in this scenario Chronos and Astaire pods can be terminated, so long as this is done gracefully such that their prestop event hook is executed.  This means that Astaire and Chronos clusters (under Kubernetes) can be dynamcially scaled up and down.
```
Comment (johadalin) [inline README.md:46]: Worth mentioning that you shouldn't do both at the same time? or in quick succession?

File: README.md [hunk 4]
```diff
@@ -159,6 +160,14 @@ If you have had to expose Bono and Ellis in a non-standard manner, you may need
 rake test[default.svc.cluster.local] PROXY={{Bono external IP addresss}} ELLIS={{external IP address of one of your nodes}}:30080 SIGNUP_CODE=secret
```
Comment (johadalin) [inline README.md]: As we're now setting everything up as headless services, we should update this to something like `rake test[default.svc.cluster.local] PROXY={{Bono service DNS/IP}} ELLIS={{Ellis service/IP}} SIGNUP_CODE=secret`

File: README.md [hunk 5]
```diff
@@ -159,6 +160,14 @@ If you have had to expose Bono and Ellis in a non-standard manner, you may need
...
 ```
 
+### Scaling the deployment
+
+All of the stateless Clearwater services: i.e. everything except Cassandra, Chronos and Astaire can be dynamically scaled up and down by running e.g. 
+`kubectl scale deployment sprout --replicas=3`
+(The other exception is Bono depending on how it is exposed external to your deployment -- see above).
```
Comment (johadalin) [inline README.md]: We're saying "(see limitations above)" way too many times here. Just needs a bit of rewording for readability. 
"Other than the limitations stated in the 'limitations' section, all our stateless services can be scaled up and down; For stateful services, there are some further restrictions, but in general all can be scaled up, and only cassandra can't be scaled down."?

File: base/pre-stop [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+supervisorctl stop all
+/etc/init.d/clearwater-etcd decommission
+
```
Comment (johadalin) [inline base/pre-stop]: blank line

Comment (johadalin) [reply base/pre-stop]: Also, i assume this probably needs to have the right permissions, but i can't see that here. I'm guessing all is good though as you've tested this live.

File: k8s-subst [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+# This script generates K8s yaml files from a set of templates.  To run it you need to 
+# define env vars IMAGE_PATH and IMAGE_TAG such that clearwater images can be found at
```
Comment (johadalin) [inline k8s-subst]: Why have we gone with having to define these as env vars, rather than passing them in as arguments to the script?
I'm not sure which way i prefer, but having them as env vars feels like it  might carry some  risk of contamination? Or possibly of people assuming they can then later just change the env vars, without needing to remodify the yaml files

File: k8s-subst [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#   IMAGE_PATH/{container name e.g. bono}:IMAGE_TAG
+# The script takes a single parameter which is the path to the Kubernetes directory 
+# (which has the templates subdirectory).
+
+[ -z "$IMAGE_PATH" ] && { echo "Need to set IMAGE_PATH"; exit 1; }
+[ -z "$IMAGE_TAG" ] && { echo "Need to set IMAGE_TAG"; exit 1; }
+[ -z "$1" ] && { echo "Need to specify path to K8s tmpl files"; exit 1; }
```
Comment (johadalin) [inline k8s-subst]: It's actually the path to the directory with the templates dir in it, right? If i passed in kubernbetes/templates/ this would fail, despite that being the path of the tmpl files.
Message should be clear what it's requiring, and possibly include an example. 
Equally, do we actually even need this? Can't the script just work relative to its location and 'know' where the templates are without user input. 
It would mean you can't run it over different templates as easily, but i don't think we need/want that flexibility. Should just be a 'run this script here' thing

File: k8s-subst [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+TMPL_PATH=$1/templates/*
+
+for f in $TMPL_PATH
+do
+  filename=${f##*/}
+  containername="${filename%.*}" 
+  sed 's@{{IMAGE:\([^}]*\)}}@'$IMAGE_PATH'\/\1\:'$IMAGE_TAG'@g' $1/templates/$containername.tmpl > $1/$containername.yaml
```
Comment (johadalin) [inline k8s-subst]: Bash is terrible. These three lines are just unreadable. 
I'm going to assume they work as expected and tested, because you've used them, but ugh. Bash. 

I don't know if it's quite right to call it 'containername' though, given it's actually going to be e.g. sprout-svc. sprout-depl. 
Not sure how much i care, but would be good if it's possible to make this section more readable? Possibly just a comment on what the steps are actually doing.

File: kubernetes/k8s-check_cluster_state [hunk 10]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (johadalin) [inline kubernetes/k8s-check_cluster_state:1]: Similarly, a comment in these scripts would be nice, though somewhat less necessary.
This is "Runs cw-check_cluster_state in the first pod of any of the astaire, cassandra, or chronos services"??

File: kubernetes/k8s-connect [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Comment (johadalin) [inline kubernetes/k8s-connect:1]: Comment please.
"Runs bash in the first pod for a given service"?

File: kubernetes/templates/astaire-depl.tmpl [hunk 12]
```diff
@@ -28,13 +29,17 @@ spec:
         livenessProbe:
           tcpSocket:
             port: 11311
-          initialDelaySeconds: 30
+          initialDelaySeconds: 90
```
Comment (johadalin) [inline kubernetes/templates/astaire-depl.tmpl]: Is this to avoid the fact that astaire containers often hit a restart on first boot? 
Seems like quite a long wait, given you were annoyed at things taking a minute or more to spin up before

File: kubernetes/templates/chronos-depl.tmpl [hunk 13]
```diff
@@ -29,13 +30,17 @@ spec:
         livenessProbe:
           tcpSocket:
             port: 7253
-          initialDelaySeconds: 30
+          initialDelaySeconds: 90
```
Comment (johadalin) [inline kubernetes/templates/chronos-depl.tmpl]: As above, why such an increase here?

File: k8s-subst (hunk 9 above)
Comment (johadalin) [inline k8s-subst]: Coming back to this after looking below, i think it would be much nicer to do this in separate steps.
in the depl files you've got {{IMAGE NAME:<process>}}, and you've got some messy regex stuff here to pull that out, and include it.
Could we not do this by having the template laid out as `{{IMAGE NAME}}:<process>{{IMAGE TAG}}`, and then in this script:

* Copy .tmpl out to .yaml
* Sed to replace {{IMAGE NAME}}
* Sed to replace/remove {{IMAGE TAG}}

possibly we could simplify the filename bit as well, but i can't work out quite how, and i'm not too worried if we can split the sed command, and add some comments on what each step is doing in a human readable form.

**Review by johadalin — COMMENTED** [review]
Mostly, i think this is all good, and a nice way to structure it all going forward. I've made a number of comments that are all fairly insignificant stuff, on comments and stuff.

I think the big thing is in the k8s-subst script; the single messy sed line is hard to follow, and  i think we could re-structure how its working to make it easier to follow, and probably easier to manage in future. Let me know if any of my comments don't make sense :)

File: README.md (hunk 3 above)
Comment (MatMeredith) [reply README.md:46]: Add below.

File: kubernetes/templates/astaire-depl.tmpl (hunk 12 above)
Comment (MatMeredith) [reply kubernetes/templates/astaire-depl.tmpl]: Yes.   The container doesn't report that it is alive until it has clustered I think and if clustering takes too long then we kill it off and restart it which can lead to the cluster ending up broken.   After conversation yesterday  I've now increased the number of retries instead.

**Comment by MatMeredith** [comment]
Sorry -- that was a bit rubbish.   How's this?

File: kubernetes/k8s-gencfg [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+"""
+Generate K8s config from template files
+"""
+
+import argparse
+import os
+import re
+import subprocess
+
+def parse_file(args, input_file, output_file):
```
Comment (johadalin) [inline kubernetes/k8s-gencfg]: I'd argue this does more than just parse, more modifies. Don't really care though

**Review by johadalin — COMMENTED** [review]
I like these updates much more. All much more readable, and followable :+1:
One comment on a function name, but really not important enough to change i think.
